text/0003-content_addressed_file_system.md

```diff
@@ -0,0 +1,143 @@
+starting with a proof of concept based on IPFS and an in-memory implementation. 
+Over time we'll work to add additional stores, which will undoubtably affect 
+the overall interface definition.
+
+A tacit goal of this interface is to manage the seam between graph-based
+storage systems, and and a file interface.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are two key interfaces to CAFS, the rest build upon these two:
+
+### File
+File is an interface based largely on `os.File` interface from golang `os` 
+package, with the exception that files can be _either a file or a directory_.
+This file interface will have many dependants in the qri ecosystem.
+
+### Filestore
+Filestore is the interface for storing files & directories. The "content 
+addressed" part means that `Put` operations are in charge of returning the name
+of the file.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The File Interface:
+```golang
+  // File is an interface that provides functionality for handling
+  // files/directories as values that can be supplied to commands. For
+  // directories, child files are accessed serially by calling `NextFile()`.
+  type File interface {
+    // Files implement ReadCloser, but can only be read from or closed if
+    // they are not directories
+    io.ReadCloser
+
+    // FileName returns a filename associated with this file
+    FileName() string
+
+    // FullPath returns the full path used when adding this file
+    FullPath() string
+
+    // IsDirectory returns true if the File is a directory (and therefore
+    // supports calling `NextFile`) and false if the File is a normal file
+    // (and therefor supports calling `Read` and `Close`)
+    IsDirectory() bool
+
+    // NextFile returns the next child file available (if the File is a
+    // directory). It will return (nil, io.EOF) if no more files are
+    // available. If the file is a regular file (not a directory), NextFile
+    // will return a non-nil error.
+    NextFile() (File, error)
+  }
+```
+
+The Filestore interface:
+```golang
+
+  // Filestore is an interface for working with a content-addressed file system.
+  // This interface is under active development, expect it to change lots.
+  // It's currently form-fitting around IPFS (ipfs.io), with far-off plans to generalize
+  // toward compatibility with git (git-scm.com), then maybe other stuff, who knows.
+  type Filestore interface {
+    // Put places a file or a directory in the store.
+    // The most notable difference from a standard file store is the store itself determines
+    // the resulting key (google "content addressing" for more info ;)
+    // keys returned by put must be prefixed with the PathPrefix,
+    // eg. /ipfs/QmZ3KfGaSrb3cnTriJbddCzG7hwQi2j6km7Xe7hVpnsW5S
+    Put(file File, pin bool) (key datastore.Key, err error)
```

Comment:
We never define `datastore`. Mention at least a stub definition up above.

`pin` isn't mentioned here either. Should have at least mention a quick explanation of what the intent is.

Reply:
Ah nice catch. Regarding "datastore", I'm going to change that to raw string values instead, b/c I don't think we need to talk about the datastore interface at all here (I think it'll only add confusion, and is specific to IPFS).

Re pin, will def add a comment